### ISSUE TITLE
Empty test valid?

### ISSUE THREAD

**Issue by Herst** [koalaman/shellcheck#459]
I was trying out the tests from http://tldp.org/LDP/abs/html/testconstructs.html and I got a parser error for the following part:

``` bash
echo "Testing \"NULL\""
if [ ]        # NULL (empty condition)
then
  echo "NULL is true."
else
  echo "NULL is false."
fi            # NULL is false.
```

Now it might not make much sense in the real world, but it seems like empty test expression is valid too (returning a non-true value). I don't know what any standards say though.

**Comment by kurahaupo** [comment]
TLDP is, shall I say politely, _not_ recommended for guidance on writing shell scripts. (Which is problematic, since that's how it bills itself.) It ignores most corner cases and proffers examples that will fail in tricky-to-detect ways when given something as simple as a filename with a space in it.

So in short, an empty test was always a bug, because it means someone has written ```[ $var ]``` instead of ```[ -n "$var" ]```.

Of course, a syntactically invalid test returns a non-zero status, which is treated as false; the difference is it may also emit a diagnostic.

**Comment by Herst** [comment]
> because it means someone has written [ $var ] instead of [ -n "$var" ].

But it wouldn't warn about the following:
```bash
#!/bin/sh
var=true
if [ $var ]; then
  echo foo
fi

```

**Comment by deryni** [comment]
Yeah, the ABS is ... poor ... at best.

An empty `test` expression, while pointless, is valid per spec [test](http://pubs.opengroup.org/onlinepubs/9699919799/utilities/test.html):

> The algorithm for determining the precedence of the operators and the return value that shall be generated is based on the number of arguments presented to test. (However, when using the "[...]" form, the <right-square-bracket> final argument shall not be counted in this algorithm.)

> In the following list, $1, $2, $3, and $4 represent the arguments presented to test:

> 0 arguments:
&nbsp;&nbsp;&nbsp;&nbsp;Exit false (1).

So shellcheck should probably parse it correctly.

**Comment by arth1** [comment]
It's not unheard of to do eternal loops this way:

until test; do
 ...
done

I'd rather empty tests be left alone, as they're usually intentional.

**Comment by deryni** [comment]
while :; do
    ....
    done

(or `while true; do` or `until false; do`) seems like a much better idea than `until test; do` but that's not really related to the issue here which is about shellcheck not correctly parsing `[ ]` as a valid test expression.

**Comment by koalaman** [comment]
This has been fixed in 5c01b6c7f. They should now parse correctly.

I agree that `false` is a way more descriptive way of doing it, so there's a style suggestion for that. An empty condition in an arithmetic loop `for ((;;))` is always true, but an empty `[ ]` is always false. Might as well be explicit.
